libsolidity/ast/AST.h

```diff
@@ -2084,7 +2084,10 @@ class UnaryOperation: public Expression
 		m_subExpression(std::move(_subExpression)),
 		m_isPrefix(_isPrefix)
 	{
-		solAssert(TokenTraits::isUnaryOp(_operator), "");
+		// NOTE: Unary plus is disallowed but only at the analysis stage.
+		// If you use `--stop-after parsing`, you can still get an AST with UnaryOperation node
+		// representing the unary plus.
+		solAssert(TokenTraits::isUnaryOp(_operator) || _operator == Token::Add);
```

Comment:
Having a binary operator in the `UnaryOperation` AST node is weird. The alternative would be to reject it already at the parsing stage but I think that would be a breaking change. If we want that, I could make the PR target `breaking`.

Reply:
I think this is a good idea.
As additional information, I noticed that there are some places in the code that still account for unary `+`:
 - `FixedPointType::unaryOperatorResult` (which is not a type that can be used, just declared):
 https://github.com/ethereum/solidity/blob/02e936ad8295c0fbed4197182eec4cac997a5048/libsolidity/ast/Types.cpp#L827-L832
 - `ExpressionCompiler::visit(UnaryOperation)`:
https://github.com/ethereum/solidity/blob/02e936ad8295c0fbed4197182eec4cac997a5048/libsolidity/codegen/ExpressionCompiler.cpp#L513-L515
 - `IRGeneratorForStatements::visit(UnaryOperation)` has an assertion for `+`, which maybe should be the case for the previous occurrences
 https://github.com/ethereum/solidity/blob/02e936ad8295c0fbed4197182eec4cac997a5048/libsolidity/codegen/ir/IRGeneratorForStatements.cpp#L762-L764
 - Also, `SMTEncoder::endVisit(UnaryOperation)` seems to accept unary `+` too
https://github.com/ethereum/solidity/blob/02e936ad8295c0fbed4197182eec4cac997a5048/libsolidity/formal/SMTEncoder.cpp#L490

Reply:
We don't have to consider moving an error that always happens in analysis to the parsing stage as breaking.
We're also likely to abandon the current breaking branch, so there's no point in targetting it :-).
Just make it a parser error and target develop, it's fine.

Reply:
That is assuming that all cases of unary plus actually did error out during analysis - that's the case right? I.e. there is no valid code (as in code passing analysis) that contains an unary plus?

Reply:
@ekpyron

> Just make it a parser error and target develop, it's fine.

ok, even better. That's actually what I did at first, but then I changed it to a non-breaking version thinking it would be a problem. I'm pushing that older version now.

> That is assuming that all cases of unary plus actually did error out during analysis - that's the case right? I.e. there is no valid code (as in code passing analysis) that contains an unary plus?

That's the case as far as I can tell. We have no tests that would expect it to pass analysis. It definitely can't pass it as a part of `UnaryOperation` node because we have an explicit check against that. The only other situation I can think of is something in scientific notation like `123e+2` but that works fine on top of this PR (it's parsed as a part of the literal).

Reply:
I had to create a new PR. We can continue this thread there: #14139.